dbms/src/Flash/Mpp/MPPTunnelSetWriter.cpp

```diff
@@ -30,52 +30,159 @@ void checkPacketSize(size_t size)
+            break;                                                                          \
+        }                                                                                   \
+        default:                                                                            \
+            break;                                                                          \
+        }                                                                                   \
+    } while (false)
+
+static inline void updatePartitionWriterMetrics(CompressionMethod method, size_t original_size, size_t actual_size, bool is_local)
 {
-    // statistic
-    GET_METRIC(tiflash_exchange_data_bytes, type_hash_original).Increment(packet_bytes);
-    // compression method is always NONE
-    if (is_local)
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_local).Increment(packet_bytes);
-    else
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_remote).Increment(packet_bytes);
+    UPDATE_EXCHANGE_MATRIC(hash, method, original_size, actual_size, is_local);
 }
 
-void updatePartitionWriterMetrics(CompressionMethod method, size_t original_size, size_t sz, bool is_local)
+template <bool is_broadcast, typename FuncIsLocalTunnel, typename FuncWriteToTunnel>
+static void broadcastOrPassThroughWriteImpl(
+    const size_t tunnel_cnt,
+    const size_t local_tunnel_cnt, // can be 0 for PassThrough writer
+    size_t ori_packet_bytes, // original data packet size
+    TrackedMppDataPacketPtr && local_tracked_packet, // can be NULL if there is no local tunnel
+    TrackedMppDataPacketPtr && remote_tracked_packet, // can be NULL if all tunnels are local mode
+    CompressionMethod compression_method,
+    FuncIsLocalTunnel && isLocalTunnel,
+    FuncWriteToTunnel && writeToTunnel)
 {
-    // statistic
-    GET_METRIC(tiflash_exchange_data_bytes, type_hash_original).Increment(original_size);
+    assert(ori_packet_bytes > 0);
 
-    switch (method)
+    const size_t remote_tunnel_cnt = tunnel_cnt - local_tunnel_cnt;
+    auto remote_tracked_packet_bytes = remote_tracked_packet ? remote_tracked_packet->getPacket().ByteSizeLong() : 0;
+
+    if (!local_tracked_packet)
     {
-    case CompressionMethod::NONE:
+        assert(local_tunnel_cnt == 0);
+        assert(remote_tracked_packet);
+    }
+    else
     {
-        if (is_local)
+        checkPacketSize(ori_packet_bytes);
+    }
+
+    if (!remote_tracked_packet)
+    {
+        assert(local_tracked_packet);
+        assert(local_tunnel_cnt == tunnel_cnt);
+    }
+    else
+    {
+        checkPacketSize(remote_tracked_packet_bytes);
+    }
+
+    // TODO avoid copy packet for broadcast.
+    for (size_t i = 0, local_cnt = 0, remote_cnt = 0; i < tunnel_cnt; ++i)
+    {
+        if (isLocalTunnel(i))
         {
-            GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_local).Increment(sz);
+            local_cnt++;
```

Comment:
```suggestion
            ++local_cnt;
```

Reply:
It's also an optional way, but there is no different for modern compiler.